Add std.meta.Stride

wilzbach opened the pull request:
As suggested in  https://github.com/dlang/phobos/pull/5596#discussion_r126774763 this adds `Stride` to std.meta.
`Stride` allows selecting a stride of the template argument list, e.g.:

```d
alias attribs = AliasSeq!(short, int, long, ushort, uint, ulong);
alias attribs = AliasSeq!(short, int, long, ushort, uint, ulong);
static assert(is(Stride!(3, attribs) == AliasSeq!(short, ushort)));
static assert(is(Stride!(3, attribs[1 .. $]) == AliasSeq!(int, uint)));
static assert(is(Stride!(-3, attribs[1 .. $]) == AliasSeq!(ulong, long)));
```

wilzbach commented on std/meta.d:
```diff
@@ -1620,3 +1621,36 @@ private template Pack(T...)
+Selects a subset of the argument list by stepping with fixed stepSize over the list.
+Params:
+    stepSize = Number of elements to increment on each iteration. Needs to be greater than `0`.
+    offset = Initial offset
```
I am not sure about `offset`, as it's basically the same as `Args[offset .. $]`

MetaLang replied:
I'd suggest removing this parameter and adding a documented example showing how to use an offset.

Biotronic replied:
Agreed. Using `Args[offset..$]` is not significantly more complicated, and the use of two consecutive `size_ts` practically guarantees that devs and reviewers will need to look up the argument order on every use of the function.

wilzbach commented:
As for unknown reasons the `CODEOWNERS` file doesn't work yet (GH bug?), I am CCing our `std.meta` experts manually:  @Biotronic @klickverbot @MetaLang @ZombineDev

MetaLang commented on std/meta.d:
```diff
@@ -1620,3 +1621,36 @@ private template Pack(T...)
+Returns: A template argument list filtered by the selected stride.
+*/
+template Stride(size_t stepSize, size_t offset, Args...)
+if (stepSize > 0)
```
Do you think it's worth allowing a negative stride that iterates from the end of the list?

Biotronic commented on std/meta.d:
```diff
@@ -1620,3 +1621,36 @@ private template Pack(T...)
+    else static if (stepSize >= Args.length)
+        alias Stride = AliasSeq!(Args[offset]);
+    else
+        alias Stride = AliasSeq!(Args[offset], Stride!(stride, offset, Args[stepSize .. $]));
```
`stride` should probably be `stepSize` here, no?

Biotronic commented on std/meta.d:
```diff
@@ -1620,3 +1621,36 @@ private template Pack(T...)
+    Args = Template arguments
+Returns: A template argument list filtered by the selected stride.
+*/
+template Stride(size_t stepSize, size_t offset, Args...)
```
Is there a good reason to use `size_t` instead of `int` for `stepSize` and `offset`? Currently, `Stride!(-1, 0, short, int, long)` compiles and violates the principle of least astonishment.

wilzbach commented:
>>> I am not sure about `offset`, as it's basically the same as `Args[offset .. $]`

>> I'd suggest removing this parameter and adding a documented example showing how to use an offset.

> Agreed. Using `Args[offset..$]` is not significantly more complicated, and the use of two consecutive `size_ts` practically guarantees that devs and reviewers will need to look up the argument order on every use of the function.

Excellent that we are all in agreement -> removed.

> Is there a good reason to use `size_t` instead of `int` for `stepSize` and `offset`? Currently, `Stride!(-1, 0, short, int, long)` compiles and violates the principle of least astonishment.

`size_t` was used in the codebase where this was taken from, but I agree with you ;-)

> Do you think it's worth allowing a negative stride that iterates from the end of the list?

Sounds interesting. Why not?

Biotronic commented on std/meta.d:
```diff
@@ -1620,3 +1621,52 @@ private template Pack(T...)
+    static assert(Pack!(Stride!(3, attribs[1 .. $])).equals!(int, uint));
+    static assert(Pack!(Stride!(-3, attribs)).equals!(ulong, long));
+
+    static assert(!__traits(compiles, Stride!(0, int)));
```
Missing test cases: `abs(stepSize) > Args.length`.

wilzbach commented:
Regarding the CircleCi failures - this was due to a `private:` label in `std.meta` and thus `Pack` and `Stride` being private. I moved `Stride` up, changed the public tests (and changelog) to use `is` and squashed everything.

andralex commented on std/meta.d:
```diff
@@ -1494,6 +1495,59 @@ template staticIsSorted(alias cmp, Seq...)
 
+/**
+Selects a subset of the argument list by stepping with fixed `stepSize` over the list.
+A negative `stepSize` starts iteration at the end of the list.
```
s/at the end of the list/with the last list element/

wilzbach commented:
> andralex approved these changes 7 hours ago
> wilzbach requested review from klickverbot and ZombineDev as code owners 12 minutes ago

Argh - apparently the CODEOWNERS feature from GitHub isn't really smart :/

wilzbach commented:
Closing and re-opening due to this weird CircleCi behavior:

![image](https://user-images.githubusercontent.com/4370550/28161539-27ebbe0a-67c4-11e7-8b5b-c7f47959d563.png)
